transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,307 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsfrag
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+)
+
+// TLS record layout from [RFC 8446]:
+//
+//	+-------------+ 0
+//	| ContentType |
+//	+-------------+ 1
+//	|  Protocol   |
+//	|  Version    |
+//	+-------------+ 3
+//	|   Record    |
+//	|   Length    |
+//	+-------------+ 5
+//	|    Data     |
+//	|     ...     |
+//	+-------------+ Record Length + 5
+//
+//	ContentType := invalid(0) | handshake(22) | application_data(23) | ...
+//	Protocol Version (deprecated) := 0x0301 ("TLS 1.0") | 0x0303 ("TLS 1.2" & "TLS 1.3") | 0x0302 ("TLS 1.1")
+//	0 < Record Length (of handshake)        ≤ 2^14
+//	0 ≤ Record Length (of application_data) ≤ 2^14
+//
+// [RFC 8446]: https://datatracker.ietf.org/doc/html/rfc8446#section-5.1
+const (
+	tlsRecordWithTypeSize          = 1 // the minimum size that contains record type
+	tlsRecordWithVersionHeaderSize = 3 // the minimum size that contains protocol version
+	tlsRecordHeaderSize            = 5 // the minimum size that contains the entire header
+	tlsTypeHandshake               = 22
+	tlsMaxRecordLen                = 1 << 14
+)
+
+// errInvalidTLSClientHello is the error used when the data received is not a valid TLS Client Hello.
+// Please use [errors.Is] to compare the returned err object with this instance.
+var errInvalidTLSClientHello = errors.New("not a valid TLS Client Hello packet")
+
+func isTLSRecordTypeHandshake(hdr []byte) bool {
```

Comment:
We can use the `_ = hdr[x]` to check instead of returning an `error` object, similar to the `binary` package implementation mentioned below.

Reply:
Right, I'm aware of that. `panic` is the expected behavior. The early check is to prevent putting garbage data into the buffer if it's not big enough.